miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/service/RunScannerClient.java

```diff
@@ -110,16 +110,48 @@ private void processResults(List<NotificationDto> results) {
           Predicate<SequencingParameters> isMatchingSequencingParameters;
           switch (dto.getPlatformType()) {
           case ILLUMINA:
-            isMatchingSequencingParameters = params -> params.getInstrumentModel().getPlatformType() == PlatformType.ILLUMINA &&
-                Math.abs(params.getReadLength() - ((IlluminaNotificationDto) dto).getReadLength()) < 2
-                && params.isPaired() == ((IlluminaNotificationDto) dto).isPairedEndRun()
-                && params.getChemistry() == Dtos.getMisoIlluminaChemistryFromRunscanner(((IlluminaNotificationDto) dto).getChemistry());
+            isMatchingSequencingParameters = params -> {
+              IlluminaNotificationDto illuminaDto = (IlluminaNotificationDto) dto;
+              if (params.getInstrumentModel().getPlatformType() != PlatformType.ILLUMINA) {
+                return false;
+              }
+              if (params.getChemistry() != Dtos.getMisoIlluminaChemistryFromRunscanner(illuminaDto.getChemistry())) {
+                return false;
+              }
+              // If we are talking to an old Run Scanner that doens't provide read lengths, use the old logic
+              if (illuminaDto.getReadLengths() == null) {
+                // The read length must match the first read length
+                if (Math.abs(params.getReadLength() - illuminaDto.getReadLength()) < 2) {
+                  // if there is no second read length, then this must be single ended
+                  if (params.getReadLength2() == 0) {
+                    return !illuminaDto.isPairedEndRun();
+                  } else {
+                    // If there is, it must be paired and symmetric
+                    return illuminaDto.isPairedEndRun() && params.getReadLength() == params.getReadLength2();
+                  }
+                } else {
+                  return false;
+                }
+              }
+              // If we have real read lengths, check they match what we see from Run Scanner
+              if (illuminaDto.getReadLengths().size() == 0) {
+                return false;
+              }
+              if (Math.abs(params.getReadLength() - illuminaDto.getReadLengths().get(0)) > 1) {
+                return false;
+              }
+              // If no second read is required, make sure none is provided
+              if (params.getReadLength2() == 0) {
+                return illuminaDto.getReadLengths().size() == 1;
+              }
+              // Otherwise, check the second read matches the right length
+              return Math.abs(params.getReadLength2() - illuminaDto.getReadLengths().get(1)) < 2;
```

Comment:
``
2019-09-09 10:03:56 ERROR: Failed to save run: 140103_D00353_0031_BC3E4NACXX
java.lang.IndexOutOfBoundsException: Index: 1, Size: 1
  at java.util.ArrayList.rangeCheck(ArrayList.java:657)
  at java.util.ArrayList.get(ArrayList.java:433)
  at uk.ac.bbsrc.tgac.miso.webapp.service.RunScannerClient.lambda$processResults$0(RunScannerClient.java:148)
```

Reply:
Fixed. I had a condition backwards.